File: families/llama/tests/test_speculative_contract.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

import pytest
from dataclasses import replace

from ..speculative.contract import EngineContract, ExecutionProfile, tree_visibility


def test_siblings_are_hidden_and_ancestors_visible():
    mask = tree_visibility([-1, 0, 0, 1, 2])
    assert mask[3] == [1, 1, 0, 1, 0]
    assert mask[4] == [1, 0, 1, 0, 1]


@pytest.mark.parametrize("parents", [[], [0], [-1, 1], [-1, -1], [-1, 2, 0]])
def test_invalid_tree_rejected(parents):
    with pytest.raises(ValueError):
        tree_visibility(parents)


def test_contract_rejects_incompatible_precision_and_capacity():
    fields = dict(role="target", layers=32, hidden_size=4096, kv_heads=8,
                  head_dim=128, vocab_size=128256, capacity=2048, feature_width=12288)
    with pytest.raises(ValueError):
        EngineContract(**fields, precision="bf16")
    with pytest.raises(ValueError):
        EngineContract(**fields, max_query=2049)
    with pytest.raises(ValueError, match="at least two"):
        EngineContract(**{**fields, "role": "draft", "vocab_size": 1})


def test_contract_rejects_unsupported_state_semantics():
    contract = EngineContract("target", 32, 4096, 8, 128, 128256, 2048, feature_width=12288)
    for changes in ({"version": 2}, {"cache_layout": "unknown"},
                    {"cache_update": "unknown"}, {"alias_contract": "unknown"}):
        with pytest.raises(ValueError):
            replace(contract, **changes)


def test_phase_profiles_separate_query_and_selected_rows():
    profiles = (ExecutionProfile("prefill", (1, 1024, 1024), (1, 1, 1)),
                ExecutionProfile("decode", (1, 5, 9), (1, 5, 9)))
    contract = EngineContract("target", 32, 4096, 8, 128, 128256, 2048,
                              max_query=1024, feature_width=12288, execution_profiles=profiles)
    assert contract.profile_shapes("attention_mask", (-1, 2048)) == [
        ((1, 2048), (1024, 2048), (1024, 2048)), ((1, 2048), (5, 2048), (9, 2048))]
    assert contract.profile_shapes("logits_indices", (-1,)) == [
        ((1,), (1,), (1,)), ((1,), (5,), (9,))]
    assert contract.profile_shapes("token_id", (-1,))[1] == ((1,), (5,), (9,))
    for changes in ({"execution_profiles": profiles[::-1]}, {"max_query": 2048},
                    {"role": "draft"}, {"execution_profiles": profiles[:1]}):
        with pytest.raises(ValueError):
            replace(contract, **changes)
    legacy = replace(contract, execution_profiles=(), max_query=64)
    assert legacy.profile_shapes("logits_indices", (-1,)) == [((1,), (16,), (64,))]


@pytest.mark.parametrize("query,logits", [((0, 5, 9), (1, 1, 1)), ((1, 9, 5), (1, 1, 1)),
                                         ((1, 5, 9), (1, 6, 9)), ((1, 5, 9), (1, 5, 10))])
def test_invalid_profile_bounds(query, logits):
    with pytest.raises(ValueError):
        ExecutionProfile("decode", query, logits)

File: families/llama/tests/cpp/test_speculative_contract.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
#include "families/llama/runtime/speculative/engine.h"

#include <iostream>
#include <stdexcept>

namespace {
template <typename F>
void expect_invalid(F&& call) {
    try {
        call();
    } catch (const std::invalid_argument&) {
        return;
    }
    throw std::runtime_error("invalid contract or result was accepted");
}
} // namespace

int main() {
    using namespace trtmc::llama::speculative;
    try {
        auto manifest = nlohmann::json::parse(R"({
          "version":1,"role":"target","precision":"fp16",
          "cache_layout":"batch_heads_capacity_dim","cache_update":"aliased_contiguous_append",
          "layers":32,"hidden_size":4096,"kv_heads":8,"head_dim":128,"vocab_size":128256,
          "capacity":2048,"max_query":1024,"feature_width":12288,
          "execution_profiles":[
            {"phase":"prefill","query":[1,1024,1024],"logits":[1,1,1]},
            {"phase":"decode","query":[1,5,9],"logits":[1,5,9]}]})");
        auto contract = Contract::parse(manifest);
        if (contract.query_limit(Phase::kPrefill) != 1024 ||
            contract.query_limit(Phase::kDecode) != 9)
            throw std::runtime_error("phase query limits are not independent");
        manifest["execution_profiles"][0]["phase"] = "decode";
        expect_invalid([&] { Contract::parse(manifest); });
        manifest["execution_profiles"][0]["phase"] = "prefill";
        manifest["execution_profiles"][1]["query"] = {1, 5, 9, 64};
        expect_invalid([&] { Contract::parse(manifest); });
        manifest.erase("execution_profiles");
        manifest["max_query"] = 64;
        contract = Contract::parse(manifest);
        if (contract.query_limit(Phase::kPrefill) != 64 ||
            contract.query_limit(Phase::kDecode) != 64)
            throw std::runtime_error("single-profile contract changed");
        manifest["version"] = 2;
        expect_invalid([&] { Contract::parse(manifest); });
        manifest["version"] = 1;
        manifest["role"] = "draft";
        manifest["vocab_size"] = 1;
        expect_invalid([&] { Contract::parse(manifest); });

        // GPU selection/path algorithms have independent oracle tests in
        // selection_gpu.py and device_policy_gpu.py. Check the readback ABI here.
        StepResult result;
        result.vocab = 4;
        result.ranks = 1;
        result.selection = {2, 1, 0, 0, 1, 1};
        if (result.token() != 2 || result.token(2) != 1)
            throw std::runtime_error("compact selection row order changed");
        expect_invalid([&] { result.token(1); }); // consumed non-finite row
        expect_invalid([&] { result.token(-1); });
        expect_invalid([&] { result.token(3); });
        expect_invalid([&] { result.token(0, 1); });
        result.ranks = 2;
        result.selection = {1, 2, 1};
        if (result.token() != 1 || result.token(0, 1) != 2)
            throw std::runtime_error("compact draft rank order changed");
        result.selection[0] = 4;
        expect_invalid([&] { result.token(); });
        result.selection[0] = -1;
        expect_invalid([&] { result.token(); });
        result.selection.clear();
        expect_invalid([&] { result.token(); });

        std::uint16_t feature_data[12]{};
        FeatureView features{feature_data, 3, 4};
        const auto last = features.slice(2, 1);
        if (last.data != feature_data + 8 || last.rows != 1 || last.width != 4)
            throw std::runtime_error("device feature row offset changed");
        expect_invalid([&] { features.slice(2, 2); });
        std::cout << "speculative contract tests passed\n";
        return 0;
    } catch (const std::exception& error) {
        std::cerr << error.what() << '\n';
        return 1;
    }
}
